### PR TITLE
Fix BOX import to not create "wrong" casts

### LINKED ISSUE #10843: Fix type of cast node generated by box import
Importing box to primitive type may produce a cast node that has "incorrect" type:
https://github.com/dotnet/coreclr/blob/ac8f8e52c49b0ffa712d68730a41883d281bea01/src/jit/importer.cpp#L5681
Normally the type of the cast node is `TYP_INT` or `TYP_LONG`, the actual type of the cast type. The above code can produce a node having small int type which is unusual. So far it seems that this is the only place where the JIT produces such a cast.

Changing the cast node type to be the actual type results in an assert here:
https://github.com/dotnet/coreclr/blob/ac8f8e52c49b0ffa712d68730a41883d281bea01/src/jit/gentree.cpp#L13240-L13242

when compiling code like:
```C#
enum ShortEnum : short
{
    None,
    Boo
}

static bool Test(ShortEnum s) => s.HasFlag(ShortEnum.Boo);
```
Those types should too be "actual" types. As is, we end up generating `TYP_SHORT` `AND`/`EQ` nodes that are also unusual and some less than ideal assembly code:
```asm
       480FBFC1             movsx    rax, cx
       BA01000000           mov      edx, 1
       6623C2               and      ax, dx
       663BC2               cmp      ax, dx
       0F94C0               sete     al
       0FB6C0               movzx    rax, al
```
Normally this should generate
```asm
       movsx    rax, cx ; this could be omitted too but that's another story
       test     al, 1
       sete     al
       movzx    rax, al
```

category:cq
theme:basic-cq
skill-level:beginner
cost:medium

### PR DESCRIPTION
**Pull request by SingleAccretion**
All casts in the IR have TYP_INT, except for this one case. Fix it by using `genActualType`.

Also add a comment explaining the reason a cast is needed at all in case the types are small.

No diffs as per SPMI. Note: it looks like the `Enum.HasFlag` optimization has been fixed to use actual types.

Fixes #10843

### CODE REVIEW

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -6855,18 +6855,26 @@ void Compiler::impImportAndPushBox(CORINFO_RESOLVED_TOKEN* pResolvedToken)
...
+            // Note regarding small types.
+            // We are going to store to the box here via an indirection, so the cast added below is
+            // redundant, since the store has an implicit truncation semantic. The reason we still
+            // add this cast is so that the code which deals with GT_BOX optimizations does not have
+            // to account for this implicit truncation (e. g. understand that BOX<byte>(0xFF + 1) is
+            // actually BOX<byte>(0) or deal with signedness mismatch and other GT_CAST complexities).
             if (srcTyp != dstTyp)
             {
-                assert((varTypeIsFloating(srcTyp) && varTypeIsFloating(dstTyp)) ||
-                       (varTypeIsIntegral(srcTyp) && varTypeIsIntegral(dstTyp)));
```
Comment (SingleAccretion) [inline src/coreclr/jit/importer.cpp:6866]: The assert was redundant with the one above the `if`.

**Comment by hughbe** [comment]
It would be good to have a test with this change

**Comment by SingleAccretion** [comment]
What kind do you have in mind? This change is not fixing a bug per se, just an inconsistency internal to the IR.

**Comment by hughbe** [comment]
I was thinking having tests that make sure `HasFlag` still produces correct results for Enums of all different sizes (unsigned/signed, 8/16/32/64 bit), e.g.
```cs
enum ShortEnum : short
{
    None,
    Boo
}

static bool Test(ShortEnum s) => s.HasFlag(ShortEnum.Boo);
```

**Comment by SingleAccretion** [comment]
I see. We already have a test for the optimization [here](https://github.com/dotnet/runtime/blob/main/src/tests/JIT/opt/Enum/hasflag.cs), though it deals more with the intricacies of how many flavors of boxes can one see. I will see about adding some more coverage of small type mixes, though I do not expect us to have issues in this area specifically (there are definitely issues with dealing with these implicit truncations in general...).

**Comment by AndyAyersMS** [comment]
@JulieLeeMSFT you can cross this one off the list I gave you the other day.

**Review by AndyAyersMS — APPROVED** [review]
LGTM.

**Pull request merged**